public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php

```diff
@@ -0,0 +1,235 @@
+		$prepared_feedback = wp_parse_args(
+			$prepared_feedback,
+			array(
+				'comment_author'       => '',
+				'comment_author_email' => '',
+				'comment_author_url'   => '',
+				'comment_author_IP'    => '',
+				'comment_agent'        => '',
+				'comment_type'         => COMMENT_TYPE, // We set this later, but it still needs to be here.
+			)
+		);
+
+		if ( ! isset( $prepared_feedback['comment_post_ID'] ) ) {
+			return new WP_Error(
+				'rest_feedback_no_post',
+				__( 'Feedback must be associated with a specific post.', 'wordcamporg' ),
+				array(
+					'status' => 400,
+				)
+			);
+		}
+
+		$feedback_author = null;
+		if ( ! empty( $prepared_feedback['user_id'] ) ) {
+			$feedback_author = absint( $prepared_feedback['user_id'] );
+		} elseif (
+			! empty( $prepared_feedback['comment_author'] )
+			&& ! empty( $prepared_feedback['comment_author_email'] )
+		) {
+			$feedback_author = array(
+				'name'  => $prepared_feedback['comment_author'],
+				'email' => $prepared_feedback['comment_author_email'],
+			);
+		} else {
+			return new WP_Error(
+				'rest_feedback_author_data_required',
+				__( 'Submitting feedback requires valid author name and email values.', 'wordcamporg' ),
+				array(
+					'status' => 400,
+				)
+			);
+		}
+
+		// We're not using the comment content field, but this also checks the length of other fields.
+		// The length of meta fields is checked separately. See `validate_feedback_meta()`.
+		$check_comment_lengths = wp_check_comment_data_max_lengths( $prepared_feedback );
+		if ( is_wp_error( $check_comment_lengths ) ) {
+			$error_code = $check_comment_lengths->get_error_code();
+			return new WP_Error(
+				$error_code,
+				__( 'Comment field exceeds maximum length allowed.', 'wordcamporg' ),
+				array(
+					'status' => 400,
+				)
+			);
+		}
+
+		// We're not using this to set the status of the feedback comment, since it's always set to `hold`. However,
+		// this also checks for duplicates, flooding, and blacklisting.
+		$allowed = wp_allow_comment( $prepared_feedback, true );
+		if ( is_wp_error( $allowed ) ) {
+			$error_code    = $allowed->get_error_code();
+			$error_message = $allowed->get_error_message();
+
+			if ( 'comment_duplicate' === $error_code ) {
+				return new WP_Error( $error_code, $error_message, array( 'status' => 409 ) );
+			}
+
+			if ( 'comment_flood' === $error_code ) {
+				return new WP_Error( $error_code, $error_message, array( 'status' => 400 ) );
+			}
+
+			return $allowed;
+		}
+
+		$meta = validate_feedback_meta( $request['meta'] ?? array() );
+		if ( is_wp_error( $meta ) ) {
+			return $meta;
+		}
+
+		$comment_id = add_feedback( $prepared_feedback['comment_post_ID'], $feedback_author, $meta );
+
+		if ( false === $comment_id ) {
+			return new WP_Error(
+				'rest_feedback_creation_failed',
+				__( 'Feedback submission failed.', 'wordcamporg' ),
+				array(
+					'status' => 400,
+				)
+			);
+		}
+
+		$response = __( 'Success!', 'wordcamporg' );
```

Comment:
Will a user ever see this? I'd expect this endpoint to only be called by javascript, so we'd want a stable return value to check against — a string that might be in a different language could make that tricky.

Reply:
Good point. It does also send a `201` response code, which should be stable. But `Success!` probably isn't necessary.

The default comment endpoint returns data about the comment on successful creation. I wasn't sure we would want to do that in this case since feedback comments are not public. What do you think would be best?

Reply:
yeah, I'm agreed on not returning comment data here, I don't think that's necessary. I didn't think about the response code, so that should be enough. Does the response need content? we might be fine with an empty response (unless that does unexpected things).

Reply:
I'll test and find out...

Reply:
Seems to work fine. I updated the response here: 3415a674